### ISSUE TITLE
[cinder-csi-plugin] Migrated Volumes get detached on kube-controller-manager restart for K8s >= 1.18.16

### ISSUE THREAD

**Issue by squ94wk** [kubernetes/cloud-provider-openstack#1531]
**Is this a BUG REPORT or FEATURE REQUEST?**:

/kind bug

**What happened**:
Short:
When restarting kube-controller-manager version 1.18.16 (e.g. after upgrade), all migrated Volumes get detached and are re-attached, but the mount is read-only.

Detailed:
A cluster is running K8s version 1.17.17 and has PVs provisioned by the in-tree Cinder plugin.
In `node.status.volumesAttached` its unique name and device path show up as:
```
kubernetes.io/cinder/de7e32ec-9fd5-4400-8d03-1c384a41971c on /dev/vdb
```
Running `lsblk` on the node gives:
```
NAME    MAJ:MIN RM  SIZE RO TYPE MOUNTPOINT
...
vdb     252:16   0    8G  0 disk /var/lib/kubelet/pods/7765675d-4316-40a9-8dde-579ddf1c0460/volumes/kubernetes.io~csi/pvc-c711bfe0-a8bd-4b7b-9f91-f38445edd248/mount
```

After enabling the CSI migration in the control plane and with fresh nodes they show up as:
```
kubernetes.io/csi/cinder.csi.openstack.org^de7e32ec-9fd5-4400-8d03-1c384a41971c on ""
```
Everything is fine.

Now kube-controller-manager is upgraded to 1.18.16 and hence restarted.
The volume gets detached from the node. It disappears from `node.status.volumesAttached` and from `lsblk`.

Now the `external-attacher` syncs and notices, that while `status.attached` in the VA is `true`, the volume is actually not attached to the node, so it re-queues it for force attachment (logs from different cluster):
```
W0428 08:32:48.377085       1 csi_handler.go:181] VA csi-51c0681ece96f23e0b0867112f498267db7ef7f2299ae9a09de035185dcd60b9 for volume 4724c0d0-d0b9-451b-b51e-ba43e2ed0c5d has attached status true but actual state false. Adding back to VA queue for forced reprocessing
```
The volume gets attached again with a different device name and without being mounted again to the *previous* mount point of the container:
```
NAME    MAJ:MIN RM  SIZE RO TYPE MOUNTPOINT
...
vdc     252:32   0    8G  0 disk
```
The mount is now read-only, because the device doesn't exist anymore (different cluster again):
```
/dev/vdd on /var/lib/kubelet/pods/552dc1bb-98a5-42e0-bcfa-5b283f17c5de/volumes/kubernetes.io~csi/pvc-5b7c6bff-1086-4cf8-9f63-3e814436dc67/mount type ext4 (ro,relatime)
```

We believe the different device name is just because of a race condition from the fast detach/attach. But the volume shouldn't be detached in the first place.

**What you expected to happen**:
No volumes get detached.

**How to reproduce it**:
- Migrate CSI on Cluster running Version 1.17.17, including nodes
- Upgrade K8s to 1.18.16 (any previous version will work fine)

**Anything else we need to know?**:
After a lot of digging, I found that the issue only happens on 1.18.16 and up.
I strongly suspect the detach is caused by this fix, that was introduced in 1.18.16 to the attacher-detacher-controller: https://github.com/kubernetes/kubernetes/pull/96617/files

The controller first builds a cache from the `node.status.volumesAttached` volume-node pairs.
With the fix it then goes through all the VAs and respective PVs and checks that `actualStateOfWorld` cache, if it's attached or detached.
It attempts to reconcile any attachments/detachments that might have gotten lost during a crash or restart.
If it can't find it attached to a node, it marks it as "Uncertain". I don't understand how this would lead to a detach though.

What I don't understand is, that the migrated PV's unique names don't match the `<plugin name>/<volume name>` format that the kube-controller-manager uses to compare. I also couldn't find where these names are generated.

Right now we can't upgrade K8s to higher than 1.18.15 and keep the PVs as they are.
We could manually replace them and keep the backing Openstack volumes, but this isn't really a solution.

**Environment**:
- cinder-csi-plugin version: v1.19.0
- OpenStack version: -
- Others:
  - Kubernetes 1.18.16
  - external csi-attacher v3.1.0

**Comment by lingxiankong** [comment]
@jichenjc @ramineni @chrigl can either of you please have a look?

**Comment by ialidzhikov** [comment]
This issue should be fixed by https://github.com/kubernetes/kubernetes/pull/101737.

/cc @Jiawei0227

**Comment by squ94wk** [comment]
This is indeed fixed upstream since 1.19.12 and 1.20.8.
